doc/load-balancing.md

```diff
@@ -41,104 +41,92 @@ It would also significantly complicate the client's code: the new design hides
-Weighted Round Robin requires that the servers report utilization to the load
-balancer in order to compute the next list of servers.
-
-## Server
-
-The gRPC Server is responsible for answering RPC requests and providing
-responses to the client. The server will also report load to the load balancer
-if a reporting stream was opened for this purpose.
+# Architecture
+
+## Overview
+
+The primary mechanism for load-balancing in gRPC is external
+load-balancing, where an external load balancer provides simple clients
+with an up-to-date list of servers.
+
+The gRPC client does support an API for built-in load balancing policies.
+However, there are only a small number of these (one of which is the
+`grpclb` policy, which implements external load balancing), and users
+are discouraged from trying to extend gRPC by adding more.  Instead, new
+load balancing policies should be implemented in external load balancers.
+
+## Workflow
+
+Load-balancing policies fit into the gRPC client workflow in between
+name resolution and the connection to the server.  Here's how it all
+works:
+
+![image](images/load-balancing.png)
+
+1. On startup, the gRPC client issues a [name resolution](naming.md) request
+   for the server name.  The name will resolve to one or more IP addresses,
+   each of which will indicate whether it is a server address or
+   a load balancer address, and a [service config](service_config.md)
+   that indicates which client-side load-balancing policy to use (e.g.,
+   `round_robin` or `grpclb`).
+2. The client instantiates the load balancing policy, which is then
+   responsible for deciding which requests will be sent to which
+   addresses.
+   - Note: If all addresses returned by the resolver are balancer
+     addresses, then the client will use the `grpclb` policy, regardless
+     of what load-balancing policy was requested by the service config.
+     Otherwise, the client will use the load-balancing policy requested
+     by the service config.  If no load-balancing policy is requested
+     by the service config, then the client will default to a policy
+     that picks the first available server address.
+3. In the case of the `grpclb` policy, it opens a stream to one of the
+   balancer addresses returned by the resolver. It asks the balancer for
+   the server addresses to use for the server name originally requested by
+   the client (i.e., the same one originally passed to the name resolver).
+   - Note: Currently, the `grpclb` policy ignores any non-balancer
+     addresses returned by the resolver. However, in the future, it may
+     be changed to use these addresses as a fallback in case no balancers
+     can be contacted.
+4. The gRPC servers to which the load balancer is directing the client
+   may report load to the load balancers, if that information is needed
+   by the load balancer's configuration.
+5. The load balancer returns a server list to the gRPC client. If the
```

Comment:
should we mention grpclb here? eg, "... to the gRPC client's grpclb policy". It may be overly verbose, but on the other hand, it may be confusing to only talk about the client (which may be interpreted as the LB addresses being visible by the application)